---
title: "AH6017 RNAseq Analysis"
author: "Hernan Lorenzi"
date: "4/13/2022"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

# Load libraries
```{r load libraries}

suppressMessages(library("RColorBrewer"))
suppressMessages(library("org.Hs.eg.db"))
suppressMessages(library("pheatmap"))
suppressMessages(library("EnhancedVolcano"))
suppressMessages(library("ggplot2"))
suppressMessages(library("ggpubr"))
suppressMessages(library("DESeq2"))
suppressMessages(library("stringr"))
suppressMessages(library("biomaRt"))
suppressMessages(library("tidyverse"))
suppressMessages(library("pcaExplorer"))
suppressMessages(library("VennDiagram"))
suppressMessages(library("clusterProfiler"))
suppressMessages(library("GOSemSim"))
suppressMessages(library("ggsci"))
suppressMessages(library("msigdbr"))
suppressMessages(library("UpSetR"))
suppressMessages(library("DOSE"))
suppressMessages(library("enrichplot"))
suppressMessages(library("ReactomePA"))
suppressMessages(library("ggupset"))
suppressMessages(library("cowplot"))
```

# Define functions
```{r aux functions}

source(file = "01_aux_rnaseq_functions.R")

```


# Load data
```{r Loading data}
all <- read.delim2("./data/TEcounts.txt", sep = "\t", header = TRUE, row.names = 1)

# Make sure read counts are numeric and rounded to 0 decimals
all.tmp <- as.data.frame(lapply(all, function(x){ round(as.numeric(x), digits = 0)} ))
rownames(all.tmp) <- rownames(all)
all <- all.tmp
rownames(all) <- str_remove(rownames(all),"[.][0-9]+$")

# Keep table with Ensemble IDs and gene Symbols
gene_symbols <- replace_gene_acc_by_symbol_ids(rownames(all))
ensembl_to_symbol <- as.data.frame(cbind("Ensembl_ID" = rownames(all), "gene_name" = gene_symbols), row.names = 1)

# Load metadata
metadata <- read.delim2("data/metadata.txt", sep = "\t", row.names = 1, header = T)

# Sort tables so metadata and read counts match order
metadata<-  metadata[match(colnames(all), rownames(metadata)), ]

# Add total read counts and sample id columns to metadata
metadata <- cbind(metadata, read_counts =colSums(all, na.rm = TRUE), sample_id = rownames(metadata))

#Remove genes with total read counts across all samples <= 20
all <- remove_all_zero_rows(all, min_total_count = 20)

# Loading gene type info from gencode.v39.primary_assembly.annotation.gtf
gene_type <- read.delim2("data/gene_type.txt", sep = "\t", row.names = 1, header = T)
rownames(gene_type) <- str_remove(rownames(gene_type),"[.][0-9]+$")
```


# Print distribution of read counts per sample
```{r}
dir.create(path = "./Plots", showWarnings = FALSE)

# Plot total reads per sample using barchar
p <- ggbarplot(data = metadata, 
          x = "sample_id", 
          y = "read_counts",
          x.text.angle = 90,
          fill = "site", 
          title = "Total read counts", 
          ylab = "Read counts",
          sort.by.groups = TRUE,
          palette = "jco",
          sort.val = "asc", 
          facet.by = "treatment")
ggsave("Plots/barplot_read_counts.pdf", plot = p)
p
```

# Normalize data to TPMs to run some comparative analysis across samples
```{r}
all.tpm <- normalize_by_TPM(all)
```

# Compare Log2FC(nuc/cyto) of treatments vs controls using TPMs
```{r}
# Keep genes with meadian TPM > 20 across all samples
all.tpm.20 <- all.tpm[rowMedians(all.tpm) > 20, ]
gene_type.20 <- gene_type[match(rownames(all.tpm.20), rownames(gene_type)),]
log2.dmso <- log(
                  rowMeans(
                    cbind(
                      all.tpm.20[,7]/all.tpm.20[,1],
                      all.tpm.20[,8]/all.tpm.20[,2],
                      all.tpm.20[,9]/all.tpm.20[,3]
                       )
                    ), 2)

log2.kpt185 <- log(
                  rowMeans(
                    cbind(
                      all.tpm.20[,10]/all.tpm.20[,4],
                      all.tpm.20[,11]/all.tpm.20[,5],
                      all.tpm.20[,12]/all.tpm.20[,6]
                       )
                    ), 2)

df <- as.data.frame(cbind(Log2dmso = log2.dmso, 
                          log2kpt185 = log2.kpt185, 
                          mean_expr = rowMeans(all.tpm.20),
                          gene_type = gene_type.20,
                          gene_id = rownames(all.tpm.20)
                          )
                    )

colnames(df) <- c("Log2dmso", "log2kpt185", "mean_expr", "gene_type", "exon_count", "gene_id")
# Set numeric fields as numeric
df$mean_expr <- as.numeric(df$mean_expr)
df$Log2dmso <- as.numeric(df$Log2dmso)
df$log2kpt185 <- as.numeric(df$log2kpt185)

# Set all pseudogene gene types as "pseudogene"
df[grep("pseudogene", df$gene_type),]$gene_type <- 'pseudogene'

# Generate scatter plot
p1 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185, color=log10(mean_expr))) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point() + geom_abline(slope = 1, intercept = 0, col = "green") + 
  geom_smooth(se = FALSE, method = "lm") + 
  theme_minimal()

p2 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185, color=gene_type)) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point() + 
  geom_smooth(se = FALSE, method = "lm") +
  geom_abline(slope = 1, intercept = 0, col = "green", size=1, linetype="dashed") +
  theme_minimal()


print(p1)
print(p2 + scale_color_brewer(palette="Spectral"))
```

# Compare Log2FC by gene type using TPMs
```{r}
gene_type_factors <- unique(df$gene_type)
for (gt in gene_type_factors){
  df.subset <- subset(df, gene_type == gt)
  
    p3 <- ggplot(df.subset, aes(x=Log2dmso, y=log2kpt185)) + 
    labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells", subtitle = paste("Gene type =",gt)) +
    geom_point( aes(fill = log2kpt185, size = round(mean_expr, digits = 0)), shape = 21, alpha = 0.7) +
    scale_fill_viridis_c() + labs(size = "Mean expr", fill = "Log2FC kpt185") +
    geom_smooth(se = TRUE, color = "darkorange",method = "lm") +
    geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size=1) +
    theme_minimal() 
    print(p3)
}
```
Non-coding RNAs that are more abundant in the nucleus compared to the cytoplasm show a drop in their relative nuclear abundance when treated with kpt185. This effect is not observed in protein-coding RNAs.

###########################################
# Analysis of expression data using DESeq2
###########################################

```{r deseq2}

# add new factor (group = treatment + site)
metadata$group <-factor(paste0(metadata$treatment,"_",metadata$site))

# Convert metadata to factors
for (variable in c("site", "time","treatment", "replicate","sample_id", "group")){
  metadata[,variable] <- as.factor(metadata[,variable]) 
}

# add new factor (group = treatment + site)
metadata$group <-factor(paste0(metadata$treatment,"_",metadata$site))

# Subset samples by treatment
all.dmso <- all[,metadata$treatment == "dmso"]
meta.dmso <- metadata[metadata$treatment == "dmso",]
meta.dmso$group <- as.factor(as.character(meta.dmso$group))
meta.dmso$treatment <- as.factor(as.character(meta.dmso$treatment))
meta.dmso$sample_id <- as.factor(as.character(meta.dmso$sample_id))

all.kpt185 <- all[,metadata$treatment == "kpt185"]
meta.kpt185 <- metadata[metadata$treatment == "kpt185",]

# Create diff expression input matrix for ALL samples
dds <- DESeqDataSetFromMatrix(countData = all, 
                              colData = metadata,  
                              design = ~ treatment + treatment:replicate + treatment:site)
dds$site <- relevel(dds$site, ref = "cytoplasmic")

# Estimate dispersion and fit model
dds <- DESeq(dds)

# Get coefs
resultsNames(dds)

# Make contrasts
# dmso nuc vs dmso cyto
res.dmso.name <- results(object = dds, name = "treatmentdmso.sitenuclear")
# kpt185 nuc vs kpt185 cyto
res.kpt185.name <- results(object = dds, name = "treatmentkpt185.sitenuclear")
# nuc/cyt effect in kpt185 vs nuc/cyt effect in dmso
res.kpt185_vs_dmso.name <- results(object = dds, contrast = list("treatmentkpt185.sitenuclear","treatmentdmso.sitenuclear"))

# Using lfcShrink instead of results to reduce high Log2FC bias of genes with low expression
# dmso nuc vs dmso cyto
res.dmso.name.ashr <- lfcShrink(dds = dds, coef = "treatmentdmso.sitenuclear", type = "ashr")
# kpt185 nuc vs kpt185 cyto
res.kpt185.name.ashr <- lfcShrink(dds = dds, coef = "treatmentkpt185.sitenuclear", type = "ashr")
# nuc/cyt effect in kpt185 vs nuc/cyt effect in dmso
res.kpt185_vs_dmso.name.ashr <- lfcShrink(dds = dds, contrast = list("treatmentkpt185.sitenuclear","treatmentdmso.sitenuclear"), type = "ashr")


# WORKING WITH ASHR RESULTS FROM NOW ON

# Add gene type info to results table
add_gene_type <- function(my_res, my_gt){
  sorted_gt <- my_gt[match(rownames(my_res), rownames(my_gt)),1]
  my_res$gene_type <- sorted_gt
  return(my_res)
}

# Add exons per gene info to results table
add_exon_count <- function(my_res, my_gt){
  sorted_gt <- my_gt[match(rownames(my_res), rownames(my_gt)),2]
  my_res$exon_count <- sorted_gt
  return(my_res)
}

# Add broader classiffication category per gene info to results table
add_broader_type <- function(my_res, my_gt){
  sorted_gt <- my_gt[match(rownames(my_res), rownames(my_gt)),3]
  my_res$broader_type <- sorted_gt
  return(my_res)
}

# Filter out DE genes with LogFC < 1 & FDR <= 0.05 
res.dmso.name.ashr.filt <- subset(res.dmso.name.ashr, padj < 0.05 & abs(log2FoldChange) >= 1)
res.kpt185.name.ashr.filt <- subset(res.kpt185.name.ashr, padj < 0.05 & abs(log2FoldChange) >= 1)
res.kpt185_vs_dmso.name.ashr.filt <- subset(res.kpt185_vs_dmso.name.ashr, padj < 0.05 & abs(log2FoldChange) >= 1)

# Add gene type column to results
res.dmso.name.ashr <- add_gene_type(res.dmso.name.ashr, gene_type)
res.kpt185.name.ashr <- add_gene_type(res.kpt185.name.ashr, gene_type)
res.kpt185_vs_dmso.name.ashr <- add_gene_type(my_res = res.kpt185_vs_dmso.name.ashr, my_gt = gene_type)

# Add exon counts column to results
res.dmso.name.ashr <- add_exon_count(res.dmso.name.ashr, gene_type)
res.kpt185.name.ashr <- add_exon_count(res.kpt185.name.ashr, gene_type)
res.kpt185_vs_dmso.name.ashr <- add_exon_count(my_res = res.kpt185_vs_dmso.name.ashr, my_gt = gene_type)

# Add broader_type column to results
res.dmso.name.ashr <- add_broader_type(res.dmso.name.ashr, gene_type)
res.kpt185.name.ashr <- add_broader_type(res.kpt185.name.ashr, gene_type)
res.kpt185_vs_dmso.name.ashr <- add_broader_type(my_res = res.kpt185_vs_dmso.name.ashr, my_gt = gene_type)

# Sort results by Log2FC, add gene symbol and save results to a txt file
res.dmso.name.ashr.logfc_sorted <- sort_and_write_res_table(res.dmso.name.ashr, "DE_dmsoNuc_vs_dmsoCyto_ashr")
res.kpt185.name.ashr.logfc_sorted <- sort_and_write_res_table(res.kpt185.name.ashr, "DE_kpt185Nuc_vs_kpt185Cyto_ashr")
res.kpt185_vs_dmso.name.ashr.logfc_sorted <- sort_and_write_res_table(res.kpt185_vs_dmso.name.ashr, "DE_kpt185_vs_dmso_ashr")

```

# Compare Log2FC
```{r}
# Sort res.kpt185.name.ashr.logfc_sorted results based on res.dmso.name.ashr.logfc_sorted order
res.kpt185.name.ashr.logfc_resorted <- res.kpt185.name.ashr.logfc_sorted[
                          match(
                            rownames(res.dmso.name.ashr.logfc_sorted), 
                            rownames(res.kpt185.name.ashr.logfc_sorted)
                            ),
                          ]
df.tmp <- as.data.frame(cbind(dmso=res.dmso.name.ashr.logfc_sorted$padj, kpt=res.kpt185.name.ashr.logfc_resorted$padj))

# Keep minimum padj value between the two conditions to be compared.
my_min_padj <- apply(df.tmp, 1, function(x) min(x, na.rm = TRUE))
df <- as.data.frame(cbind(
                          Log2dmso = res.dmso.name.ashr.logfc_sorted$log2FoldChange, 
                          log2kpt185 = res.kpt185.name.ashr.logfc_resorted$log2FoldChange, 
                          mean_expr = res.dmso.name.ashr.logfc_sorted$baseMean,
                          gene_type = res.dmso.name.ashr.logfc_sorted$gene_type,
                          broader_type = res.dmso.name.ashr.logfc_sorted$broader_type,
                          gene_id = rownames(res.dmso.name.ashr.logfc_sorted),
                          gene_name = res.dmso.name.ashr.logfc_sorted$gene_name,
                          exon_count = res.dmso.name.ashr.logfc_sorted$exon_count,
                          padj = my_min_padj
                          )
                    )

# Convert columns into numeric
for (variable in c("Log2dmso", "log2kpt185","mean_expr","padj")){
  df[,variable] <- as.numeric(df[,variable]) 
}

# Add factor for high and low Nuc/Cyt LogFC dmso ratio
df$nuc_cyt_ratio <- 'high'
df$nuc_cyt_ratio[df$Log2dmso < 0] <- 'low'

# Add kpt185/dmso ratio of corresponding Log2FC 
df$kpt185_dmso_ratio <- df$log2kpt185/df$Log2dmso

# Filter out genes with padj > 0.05
df.subset <- subset(df, padj <= 0.05)

df <- df.subset

# Set all pseudogene gene types as "pseudogene"
df[grep("pseudogene", df$gene_type),]$gene_type <- 'pseudogene'

# Generate scatter plot
p1 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185, color=log10(mean_expr))) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point() + geom_abline(slope = 1, intercept = 0, col = "green") + 
  xlab(label = "Log2FC(nuc/cyto) DMSO") + ylab(label = "Log2FC(nuc/cyto) Kpt185") +
  geom_smooth(se = TRUE, method = "lm") + 
  theme_minimal()

p2 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185 )) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point(aes(fill = gene_type), size = 2, shape = 21, alpha = 0.7) + 
  xlab(label = "Log2FC(nuc/cyto) DMSO") + ylab(label = "Log2FC(nuc/cyto) Kpt185") +
  geom_smooth(se = FALSE, method = "lm") +
  geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size = 1) +
  theme_minimal()

p3 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185 )) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point(aes(fill = broader_type), size = 2, shape = 21, alpha = 0.7) + 
  xlab(label = "Log2FC(nuc/cyto) DMSO") + ylab(label = "Log2FC(nuc/cyto) Kpt185") +
  geom_smooth(se = FALSE, method = "lm") +
  geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size = 1) +
  theme_minimal()

print(p1)
print(p2)
print(p3)

```

# Compare Log2FC by gene type using DESeq2 results
```{r}
gene_type_factors <- unique(df$gene_type)
for (gt in gene_type_factors){
  df.subset <- subset(df, gene_type == gt)
    
    p3 <- ggplot(df.subset, aes(x=Log2dmso, y=log2kpt185)) + 
    labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells", subtitle = paste("Gene type =",gt)) +
    geom_point( aes(fill = log2kpt185, size = round(mean_expr, digits = 0)), shape = 21, alpha = 0.7) +
    scale_fill_viridis_c() + labs(size = "Mean expr", fill = "Log2FC kpt185") +
    xlab(label = "Log2FC(nuc/cyto) DMSO") + ylab(label = "Log2FC(nuc/cyto) Kpt185") +
    geom_smooth(se = TRUE, color = "darkorange",method = "lm") +
    geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size = 1) +
    theme_minimal() 
    
    print(p3)
}
```

# Compare Log2FC by broader category using DESeq2 results
```{r}
gene_type_factors <- unique(df$broader_type)
for (gt in gene_type_factors){
  df.subset <- subset(df, broader_type == gt)
    
    p3 <- ggplot(df.subset, aes(x=Log2dmso, y=log2kpt185)) + 
    labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells", subtitle = paste("Gene type =",gt)) +
    geom_point( aes(fill = log2kpt185, size = round(mean_expr, digits = 0)), shape = 21, alpha = 0.7) +
    scale_fill_viridis_c() + labs(size = "Mean expr", fill = "Log2FC kpt185") +
    xlab(label = "Log2FC(nuc/cyto) DMSO") + ylab(label = "Log2FC(nuc/cyto) Kpt185") +
    geom_smooth(se = TRUE, color = "darkorange",method = "lm") +
    geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size = 1) +
    theme_minimal() 
    
    print(p3)
    ggsave2(filename = paste0(gt,"_lgFC_dmso_kpt_dotplot.pdf"), plot = p3, path = "./Plots/")
}
```
# Test if transcripts with Nuc/Cyt ratio > 0 tend to have lower Nuc/Cyt ratio after kpt185 treatment and the opposite is true for those transcripts with Nuc/Cyt ration < 0.
```{r}
gene_type_factors <- unique(df$broader_type)
for (gt in gene_type_factors){
  
    # dotplot of high and low
    
    df.subset <- subset(df, broader_type == gt)
    df.subset.test <- as.data.frame(cbind(log2FC = c(df.subset$log2kpt185, df.subset$Log2dmso),
                                          treatment = c(rep("kpt185", length(df.subset$gene_id)),
                                                        rep("dmso", length(df.subset$gene_id))
                                                        ),
                                          nuc_cyt_ratio = rep(df.subset$nuc_cyt_ratio,2)
                                          
                                          )
                                    )
    df.subset.test$log2FC <-as.numeric(df.subset.test$log2FC)
    
    my_comparisons <- list( c("kpt185", "dmso") )
    
    p5 <- ggboxplot(data = df.subset.test, 
          x = "treatment", 
          y = "log2FC", 
          color = "treatment", 
          size = 0.5,
          palette = "jco",
          facet.by = "nuc_cyt_ratio",
          title = "Comparison of  changes in subcellular distribution\nof transcript with high and low nuc/cyt ratios", 
          ylab = "kpt185/dmso Log2FC(nuc/cyt) Ratio", 
          xlab = "Treatment",
          repel = TRUE,
          ylim = c(min(df.subset.test$log2FC)-1, max(df.subset.test$log2FC)+1) ) +
          labs(title = "Comparison of  changes in transcript subcellular localization\nof transcripts with high and low nuc/cyt ratios", 
         subtitle = paste("Gene type =",gt)) + theme(legend.position="bottom") +
         stat_compare_means(method = "wilcox.test", comparisons = my_comparisons) # Add pairwise comparisons p-value
    
  print(p5)
  ggsave2(filename = paste0(gt,"_kpt_dmso_lgFC_wilx_test_boxplot.pdf"), plot = p5, path = "./Plots/")
}
```

# Run enrichment analysis for treatment effect
(res.kpt185_vs_dmso.name.ashr.logfc_sorted)
```{r gene enrichment analysis}

# Get enrichment functions
source(file = "02_Gene_enrichment_functions.R")

# Prepare gene list for the enrichment analysis
genes_up <- rownames(subset(res.kpt185_vs_dmso.name.ashr.logfc_sorted, 
                         log2FoldChange > 0 & padj <= 0.05))
genes_down <- rownames(subset(res.kpt185_vs_dmso.name.ashr.logfc_sorted, 
                         log2FoldChange < 0 & padj <= 0.05))

my_ordered_dds_res <- res.kpt185_vs_dmso.name.ashr.logfc_sorted[order(res.kpt185_vs_dmso.name.ashr.logfc_sorted$log2FoldChange, decreasing = T, na.last = T), ]
geneList <- my_ordered_dds_res$log2FoldChange
names(geneList) <- rownames(my_ordered_dds_res)


my_prefix <- "kpt185_vs_dmso"
my_prefix_up <- "kpt185_vs_dmso_UP"
my_prefix_down <- "kpt185_vs_dmso_DOWN"

  ########################################
  ## H: hallmark gene sets
  ########################################
  # Fetch gene sets
  msig_h <- msigdbr(species = "Homo sapiens", category = "H") %>% dplyr::select(gs_name, ensembl_gene, ) 

  # Run overrepresentation analysis
  run_overrepresentation_analysis(gene_set = msig_h, genes = genes_up, analysis_name = my_prefix_up, gs_name = "H", type = "NONE")
  run_overrepresentation_analysis(gene_set = msig_h, genes = genes_down, analysis_name = my_prefix_down, gs_name = "H", type = "NONE")
  
  # Run Gene Set Enrichment Analysis
  run_enrichment_analysis(gene_set = msig_h, geneList = geneList, analysis_name = my_prefix, gs_name = "H")
  
  
  ###########################################################################
  ## C3 regulatory target gene sets => TFT: transcription factor targets
  ###########################################################################
  # Fetch gene sets
  msig_c3_gtrd <- msigdbr(species = "Homo sapiens", category = "C3", subcategory = "TFT:GTRD") %>% dplyr::select(gs_name, ensembl_gene, )
  msig_c3_tft <- msigdbr(species = "Homo sapiens", category = "C3", subcategory = "TFT:TFT_Legacy") %>% dplyr::select(gs_name, ensembl_gene, )

  # Run overrepresentation analysis
  run_overrepresentation_analysis(gene_set = msig_c3_gtrd, genes = genes_up, analysis_name = my_prefix, gs_name = "C3_TFT_GTRD", type = "NONE")
  run_overrepresentation_analysis(gene_set = msig_c3_tft, genes = genes_up, analysis_name = my_prefix_up, gs_name = "C3_TFT_Legacy", type = "NONE")
  run_overrepresentation_analysis(gene_set = msig_c3_gtrd, genes = genes_down, analysis_name = my_prefix_down, gs_name = "C3_TFT_GTRD", type = "NONE")
  run_overrepresentation_analysis(gene_set = msig_c3_tft, genes = genes_down, analysis_name = my_prefix_down, gs_name = "C3_TFT_Legacy", type = "NONE")
  
  # Run Gene Set Enrichment Analysis
  run_enrichment_analysis(gene_set = msig_c3_gtrd, geneList = geneList, analysis_name = my_prefix, gs_name = "C3_TFT_GTRD")
  run_enrichment_analysis(gene_set = msig_c3_tft, geneList = geneList, analysis_name = my_prefix, gs_name = "C3_TFT_Legacy")


  ##########################################################
  ## C2 curated gene sets => CP:REACTOME Canonical pathways
  ##########################################################
  # Fetch gene sets
  msig_c2_reactome <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% dplyr::select(gs_name, ensembl_gene, )

  # Run overrepresentation analysis
  run_overrepresentation_analysis(gene_set = msig_c2_reactome, genes = genes_up, analysis_name = my_prefix_up, gs_name = "C2_CP_REACTOME", type = "reactome")
  run_overrepresentation_analysis(gene_set = msig_c2_reactome, genes = genes_down, analysis_name = my_prefix_down, gs_name = "C2_CP_REACTOME", type = "reactome")
  
  # Run Gene Set Enrichment Analysis
  run_enrichment_analysis(gene_set = msig_c2_reactome, geneList = geneList, analysis_name = my_prefix, gs_name = "C2_CP_REACTOME", type = "reactome")
```
#################################
## Gene / GO enrichment analysis
#################################

```{r GO enrichment}
# The code below selects a list of gene IDs based on a specific Log2FC cutoff for running Overrepresentation analysis with the R function 'enricher'. The list of genes are stored in the variable "gene" for gene IDs = Ensembl gene IDs, and in the variable  'gene.entrezid' for a list of Entrez gene IDs. Usually, geneset databases use eithr Entrez gene IDs or gene symbols as identifiers, so in some cases it is handy to count with a list of genes using Entrez gene IDs. 

#The same code also extract the entire list of genes used for the Differential Expression Analysis as an R list element, using Ensemble gene IDs as keys (names) and Log2FC as values. This geneList is used as input for running Gene Set Enrichment Analysis with the function 'GSEA'. If needed, the code below also generates the same geneList but using Entrez gene IDs as keys.

ggo = list()
go.enr = list()


ego_up <- enrichGO(gene = genes_up,
                  OrgDb= "org.Hs.eg.db",
                  ont = "BP", # Either MF, BP, CC or ALL
                  pAdjustMethod = "BH",
                  qvalueCutoff  = 0.05,
                  keyType = 'ENSEMBL',
                  readable      = TRUE,
                  minGSSize = 10,
                  maxGSSize = 500
                  )

ego_down <- enrichGO(gene = genes_down,
                  OrgDb= "org.Hs.eg.db",
                  ont = "BP", # Either MF, BP, CC or ALL
                  pAdjustMethod = "BH",
                  qvalueCutoff  = 0.05,
                  keyType = 'ENSEMBL',
                  readable      = TRUE,
                  minGSSize = 10,
                  maxGSSize = 500
                  )

p1_up <- dotplot(ego_up, showCategory=30)
p2_up <- mutate(ego_up, qscore = -log(p.adjust, base=10)) %>% 
    barplot(x="qscore")
p1_down <- dotplot(ego_down, showCategory=30)
p2_down <- mutate(ego_down, qscore = -log(p.adjust, base=10)) %>% 
    barplot(x="qscore")

ggsave2(filename = paste0(my_prefix,"_UP_GO_enr_BP_dotplot.pdf"), plot = p1_up, path = "./ClusterProfiler")
ggsave2(filename = paste0(my_prefix,"_UP_GO_enr_BP_barplot.pdf"), plot = p2_up, path = "./ClusterProfiler")
ggsave2(filename = paste0(my_prefix,"_DOWN_GO_enr_BP_dotplot.pdf"), plot = p1_down, path = "./ClusterProfiler")
ggsave2(filename = paste0(my_prefix,"_DOWN_GO_enr_BP_barplot.pdf"), plot = p2_down, path = "./ClusterProfiler")
onthology <- "BP"
write.table(ego_up, file = paste0("./ClusterProfiler/",my_prefix,"_",onthology,"_go_enrich_UP.txt"), sep = "\t", col.names=NA, )
write.table(ego_down, file = paste0("./ClusterProfiler/",my_prefix,"_",onthology,"_go_enrich_DOWN.txt"), sep = "\t", col.names=NA)


print(p1_up)
print(p2_up)
print(p1_down)
print(p2_down)
```



###############################################################################################################################
###############################################################################################################################
########################################################     END       ########################################################
###############################################################################################################################
###############################################################################################################################

```{r}
res.kpt185_vs_dmso.name.ashr.logfc_sorted.sig <- subset(res.kpt185_vs_dmso.name.ashr.logfc_sorted, padj <= 0.05)
```

```{r}
sessionInfo()
```