---
title: "AH6017 RNAseq Analysis"
author: "Hernan Lorenzi"
date: "4/13/2022"
output:
  html_document:
    df_print: paged
  pdf_document: default
---

# Load libraries
```{r load libraries}

suppressMessages(library("RColorBrewer"))
suppressMessages(library("org.Hs.eg.db"))
suppressMessages(library("pheatmap"))
suppressMessages(library("EnhancedVolcano"))
suppressMessages(library("ggplot2"))
suppressMessages(library("ggpubr"))
suppressMessages(library("DESeq2"))
suppressMessages(library("stringr"))
suppressMessages(library("biomaRt"))
suppressMessages(library("tidyverse"))
suppressMessages(library("pcaExplorer"))
suppressMessages(library("VennDiagram"))
suppressMessages(library("clusterProfiler"))
suppressMessages(library("GOSemSim"))
suppressMessages(library("ggsci"))
```

# Define functions
```{r aux functions}

source(file = "01_aux_rnaseq_functions.R")

```


# Load data
```{r Loading data}
all <- read.delim2("./data/TEcounts.txt", sep = "\t", header = TRUE, row.names = 1)

# Make sure read counts are numeric and rounded to 0 decimals
all.tmp <- as.data.frame(lapply(all, function(x){ round(as.numeric(x), digits = 0)} ))
rownames(all.tmp) <- rownames(all)
all <- all.tmp
rownames(all) <- str_remove(rownames(all),"[.][0-9]+$")

# Keep table with Ensemble IDs and gene Symbols
gene_symbols <- replace_gene_acc_by_symbol_ids(rownames(all))
ensembl_to_symbol <- as.data.frame(cbind("Ensembl_ID" = rownames(all), "gene_name" = gene_symbols), row.names = 1)

# Load metadata
metadata <- read.delim2("data/metadata.txt", sep = "\t", row.names = 1, header = T)

# Sort tables so metadata and read counts match order
metadata<-  metadata[match(colnames(all), rownames(metadata)), ]

# Add total read counts and sample id columns to metadata
metadata <- cbind(metadata, read_counts =colSums(all, na.rm = TRUE), sample_id = rownames(metadata))

#Remove genes with total read counts across all samples <= 20
all <- remove_all_zero_rows(all, min_total_count = 20)

# Loading gene type info from gencode.v39.primary_assembly.annotation.gtf
gene_type <- read.delim2("data/gene_type.txt", sep = "\t", row.names = 1, header = F)
rownames(gene_type) <- str_remove(rownames(gene_type),"[.][0-9]+$")
```


# Print distribution of read counts per sample
```{r}
dir.create(path = "./Plots", showWarnings = FALSE)

# Plot total reads per sample using barchar
p <- ggbarplot(data = metadata, 
          x = "sample_id", 
          y = "read_counts",
          x.text.angle = 90,
          fill = "site", 
          title = "Total read counts", 
          ylab = "Read counts",
          sort.by.groups = TRUE,
          palette = "jco",
          sort.val = "asc", 
          facet.by = "treatment")
ggsave("Plots/barplot_read_counts.pdf", plot = p)
p
```

# Normalize data to TPMs to run some comparative analysis across samples
```{r}
all.tpm <- normalize_by_TPM(all)
```

# Compare Log2FC(nuc/cyto) of treatments vs controls using TPMs
```{r}
# Keep genes with meadian TPM > 20 across all samples
all.tpm.20 <- all.tpm[rowMedians(all.tpm) > 20, ]
gene_type.20 <- gene_type[match(rownames(all.tpm.20), rownames(gene_type)),]
log2.dmso <- log(rowMeans(all.tpm.20[,7:9])/rowMeans(all.tpm.20[,1:3]), 2)
log2.kpt185 <- log(rowMeans(all.tpm.20[,10:12])/rowMeans(all.tpm.20[,4:6]), 2)
df <- as.data.frame(cbind(Log2dmso = log2.dmso, 
                          log2kpt185 = log2.kpt185, 
                          mean_expr = rowMeans(all.tpm.20),
                          gene_type = gene_type.20,
                          gene_id = rownames(all.tpm.20)
                          )
                    )

# Set numeric fields as numeric
df$mean_expr <- as.numeric(df$mean_expr)
df$Log2dmso <- as.numeric(df$Log2dmso)
df$log2kpt185 <- as.numeric(df$log2kpt185)

# Set all pseudogene gene types as "pseudogene"
df[grep("pseudogene", df$gene_type),]$gene_type <- 'pseudogene'

# Generate scatter plot
p1 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185, color=log10(mean_expr))) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point() + geom_abline(slope = 1, intercept = 0, col = "green") + 
  geom_smooth(se = FALSE) + 
  theme_minimal()

p2 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185, color=gene_type)) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point() + 
  geom_smooth(se = FALSE) +
  geom_abline(slope = 1, intercept = 0, col = "green", size=1, linetype="dashed") +
  theme_minimal()


print(p1)
print(p2 + scale_color_brewer(palette="Spectral"))
```

# Compare Log2FC by gene type using TPMs
```{r}
gene_type_factors <- unique(df$gene_type)
for (gt in gene_type_factors){
  df.subset <- subset(df, gene_type == gt)
  
    p3 <- ggplot(df.subset, aes(x=Log2dmso, y=log2kpt185)) + 
    labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells", subtitle = paste("Gene type =",gt)) +
    geom_point( aes(fill = log2kpt185, size = round(mean_expr, digits = 0)), shape = 21, alpha = 0.7) +
    scale_fill_viridis_c() + labs(size = "Mean expr", fill = "Log2FC kpt185") +
    geom_smooth(se = TRUE, color = "darkorange",method = "lm") +
    geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size=1) +
    theme_minimal() 
    print(p3)
}
```
Non-coding RNAs that are more abundant in the nucleus compared to the cytoplasm show a drop in their relative nuclear abundance when treated with kpt185. This effect is not observed in protein-coding RNAs.

###########################################
# Analysis of expression data using DESeq2
###########################################

```{r deseq2}

# add new factor (group = treatment + site)
metadata$group <-factor(paste0(metadata$treatment,"_",metadata$site))

# Convert metadata to factors
for (variable in c("site", "time","treatment", "replicate","sample_id", "group")){
  metadata[,variable] <- as.factor(metadata[,variable]) 
}

# add new factor (group = treatment + site)
metadata$group <-factor(paste0(metadata$treatment,"_",metadata$site))

# Subset samples by treatment
all.dmso <- all[,metadata$treatment == "dmso"]
meta.dmso <- metadata[metadata$treatment == "dmso",]
meta.dmso$group <- as.factor(as.character(meta.dmso$group))
meta.dmso$treatment <- as.factor(as.character(meta.dmso$treatment))
meta.dmso$sample_id <- as.factor(as.character(meta.dmso$sample_id))

all.kpt185 <- all[,metadata$treatment == "kpt185"]
meta.kpt185 <- metadata[metadata$treatment == "kpt185",]

# Create diff expression input matrix for ALL samples
dds <- DESeqDataSetFromMatrix(countData = all, 
                              colData = metadata,  
                              design = ~ treatment + treatment:replicate + treatment:site)
dds$site <- relevel(dds$site, ref = "cytoplasmic")

# Estimate dispersion and fit model
dds <- DESeq(dds)

# Get coefs
resultsNames(dds)

# Make contrasts
# dmso nuc vs dmso cyto
res.dmso.name <- results(object = dds, name = "treatmentdmso.sitenuclear")
# kpt185 nuc vs kpt185 cyto
res.kpt185.name <- results(object = dds, name = "treatmentkpt185.sitenuclear")
# nuc/cyt effect in kpt185 vs nuc/cyt effect in dmso
res.kpt185_vs_dmso.name <- results(object = dds, contrast = list("treatmentkpt185.sitenuclear","treatmentdmso.sitenuclear"))

# Using lfcShrink instead of results to reduce high Log2FC bias of genes with low expression
# dmso nuc vs dmso cyto
res.dmso.name.ashr <- lfcShrink(dds = dds, coef = "treatmentdmso.sitenuclear", type = "ashr")
# kpt185 nuc vs kpt185 cyto
res.kpt185.name.ashr <- lfcShrink(dds = dds, coef = "treatmentkpt185.sitenuclear", type = "ashr")
# nuc/cyt effect in kpt185 vs nuc/cyt effect in dmso
res.kpt185_vs_dmso.name.ashr <- lfcShrink(dds = dds, contrast = list("treatmentkpt185.sitenuclear","treatmentdmso.sitenuclear"), type = "ashr")


# WORKING WITH ASHR RESULTS FROM NOW ON

# Add gene type info to results table
add_gene_type <- function(my_res, my_gt){
  sorted_gt <- my_gt[match(rownames(my_res), rownames(my_gt)),1]
  my_res$gene_type <- sorted_gt
  return(my_res)
}

# Filter out DE genes with LogFC < 1 & FDR <= 0.05 
res.dmso.name.ashr.filt <- subset(res.dmso.name.ashr, padj < 0.05 & abs(log2FoldChange) >= 1)
res.kpt185.name.ashr.filt <- subset(res.kpt185.name.ashr, padj < 0.05 & abs(log2FoldChange) >= 1)
res.kpt185_vs_dmso.name.ashr.filt <- subset(res.kpt185_vs_dmso.name.ashr, padj < 0.05 & abs(log2FoldChange) >= 1)

# Add gene type column to results
res.dmso.name.ashr <- add_gene_type(res.dmso.name.ashr, gene_type)
res.kpt185.name.ashr <- add_gene_type(res.kpt185.name.ashr, gene_type)
res.kpt185_vs_dmso.name.ashr <- add_gene_type(my_res = res.kpt185_vs_dmso.name.ashr, my_gt = gene_type)

# Sort results by Log2FC, add gene symbol and save results to a txt file
res.dmso.name.ashr.logfc_sorted <- sort_and_write_res_table(res.dmso.name.ashr, "DE_dmsoNuc_vs_dmsoCyto_ashr")
res.kpt185.name.ashr.logfc_sorted <- sort_and_write_res_table(res.kpt185.name.ashr, "DE_kpt185Nuc_vs_kpt185Cyto_ashr")
res.kpt185_vs_dmso.name.ashr.logfc_sorted <- sort_and_write_res_table(res.kpt185_vs_dmso.name.ashr, "DE_kpt185_vs_dmso_ashr")

```

# Compare Log2FC
```{r}
# Sort res.kpt185.name.ashr.logfc_sorted results based on res.dmso.name.ashr.logfc_sorted order
res.kpt185.name.ashr.logfc_resorted <- res.kpt185.name.ashr.logfc_sorted[
                          match(
                            rownames(res.dmso.name.ashr.logfc_sorted), 
                            rownames(res.kpt185.name.ashr.logfc_sorted)
                            ),
                          ]
df.tmp <- as.data.frame(cbind(dmso=res.dmso.name.ashr.logfc_sorted$padj, kpt=res.kpt185.name.ashr.logfc_resorted$padj))
my_min_padj <- apply(df.tmp, 1, function(x) min(x, na.rm = TRUE))
df <- as.data.frame(cbind(
                          Log2dmso = res.dmso.name.ashr.logfc_sorted$log2FoldChange, 
                          log2kpt185 = res.kpt185.name.ashr.logfc_resorted$log2FoldChange, 
                          mean_expr = res.dmso.name.ashr.logfc_sorted$baseMean,
                          gene_type = res.dmso.name.ashr.logfc_sorted$gene_type,
                          gene_id = rownames(res.dmso.name.ashr.logfc_sorted),
                          gene_name = res.dmso.name.ashr.logfc_sorted$gene_name,
                          padj = my_min_padj
                          )
                    )
# Convert columns into numeric
for (variable in c("Log2dmso", "log2kpt185","mean_expr","padj")){
  df[,variable] <- as.numeric(df[,variable]) 
}

# Filter out genes with padj > 0.05
df.subset <- subset(df, padj <= 0.05)

df <- df.subset

# Set all pseudogene gene types as "pseudogene"
df[grep("pseudogene", df$gene_type),]$gene_type <- 'pseudogene'

# Generate scatter plot
p1 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185, color=log10(mean_expr))) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point() + geom_abline(slope = 1, intercept = 0, col = "green") + 
  geom_smooth(se = FALSE) + 
  theme_minimal()

p2 <- ggplot(df, aes(x=Log2dmso, y=log2kpt185, color=gene_type)) + 
  labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells") +
  geom_point() + 
  geom_smooth(se = FALSE) +
  geom_abline(slope = 1, intercept = 0, col = "green", size=1, linetype="dashed") +
  theme_minimal()


print(p1)
print(p2 + scale_color_brewer(palette="Spectral"))

```

# Compare Log2FC by gene type using DESeq2 results
```{r}
gene_type_factors <- unique(df$gene_type)
for (gt in gene_type_factors){
  df.subset <- subset(df, gene_type == gt)
  
    #p3 <- ggplot(df.subset, aes(x=Log2dmso, y=log2kpt185, color=log10(mean_expr), size = mean_expr)) + 
    #labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells", subtitle = paste("Gene type =",gt)) +
    #geom_point() + 
    #geom_smooth(se = TRUE, color = "darkorange") +
    #geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size=1) +
    #theme_minimal() 
    #print(p3)
    
    p3 <- ggplot(df.subset, aes(x=Log2dmso, y=log2kpt185)) + 
    labs(title = "Comparison of Log2FC(nuc/cyto) between kpt185 and dmso treated cells", subtitle = paste("Gene type =",gt)) +
    geom_point( aes(fill = log2kpt185, size = round(mean_expr, digits = 0)), shape = 21, alpha = 0.7) +
    scale_fill_viridis_c() + labs(size = "Mean expr", fill = "Log2FC kpt185") +
    geom_smooth(se = TRUE, color = "darkorange",method = "lm") +
    geom_abline(slope = 1, intercept = 0, col = "green", linetype="dashed", size = 1) +
    theme_minimal() 
    
    print(p3)
}
```

###############################################################################################################################
###############################################################################################################################
########################################################     END       ########################################################
###############################################################################################################################
###############################################################################################################################


```{r}
sessionInfo()
```